### PR TITLE
[Uniques V2] Feature flags

### PR DESCRIPTION
**Pull request by jsidorenko**
The PR adds a new way to configure the pallet, collection and its items.
- Pallet-level feature flags would allow disabling some functionality that is not needed in the runtime.
- For collections and items I basically extracted fields like is_frozen from the internal structures (ItemDetails, CollectionDetails, etc) and put into the new ItemConfigOf/CollectionConfigOf storage. This would allow extending the functionality of the collections/items without the need to migrate them.
- separated metadata from attributes locking. It was not obvious before that when you lock the metadata that will lock the attributes as well. Now there is an explicit method where you specify what you want to get locked

### CODE REVIEW

**Comment by gilescope** [comment]
Should we use some other term than feature flags if we're not going to implement them as rust features as that will just get confusing?

**Comment by jsidorenko** [comment]
> Should we use some other term than feature flags if we're not going to implement them as rust features as that will just get confusing?

If you check the code, we use `SystemFeatures` for pallet level features and `ItemConfig`/`CollectionConfig` for items and collections

File: frame/nfts/src/features/macros.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+					.path(Path::new("BitFlags", module_path!()))
+					.type_params(vec![TypeParameter::new("T", Some(meta_type::<$bitflag_enum>()))])
+					.composite(
+						Fields::unnamed()
+							.field(|f| f.ty::<$size>().type_name(stringify!($bitflag_enum))),
+					)
+			}
+		}
+	};
+}
```
Comment (gilescope) [inline frame/nfts/src/features/macros.rs:71]: This isn't a feature. I'd put this somewhere else. Feels like it should be in primitives somewhere as none of this is NFT related?

File: frame/nfts/src/features/settings.rs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::*;
+use frame_support::pallet_prelude::*;
+
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
```
Comment (gilescope) [inline frame/nfts/src/features/settings.rs:21]: docs here could give an example of a couple of likely nft settings so that people get the idea.

File: frame/nfts/src/functions.rs [hunk 3]
```diff
@@ -36,12 +36,20 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	) -> DispatchResult {
 		let collection_details =
 			Collection::<T, I>::get(&collection).ok_or(Error::<T, I>::UnknownCollection)?;
-		ensure!(!collection_details.is_frozen, Error::<T, I>::Frozen);
 		ensure!(!T::Locker::is_locked(collection, item), Error::<T, I>::Locked);
```
Comment (gilescope) [inline frame/nfts/src/functions.rs:39]: Would be good to split the `Locked` error into `CollectionLocked` and `ItemLocked` so that UIs can give better feedback to the user.

File: frame/nfts/src/features/macros.rs (hunk 1 above)
Comment (jsidorenko) [reply frame/nfts/src/features/macros.rs:71]: What would be the best file to put this into?

File: bin/node/runtime/src/lib.rs [hunk 4]
```diff
@@ -1497,6 +1498,10 @@ impl pallet_uniques::Config for Runtime {
 	type Locker = ();
 }
 
+parameter_types! {
+	pub FeatureFlags: SystemFeatures = SystemFeatures::empty();
```
Comment (gilescope) [inline bin/node/runtime/src/lib.rs:1502]: if someone asks me what SystemFeatures are I would say pallet features so why don't we rename `SystemFeatures` to `PalletFeatures` ?

Comment (jsidorenko) [reply bin/node/runtime/src/lib.rs]: Agree! Renamed ;)

File: frame/nfts/src/features/lock.rs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+			Ok(())
+		})
+	}
+
+	pub fn do_lock_item(
+		maybe_check_owner: Option<T::AccountId>,
+		collection: T::CollectionId,
+		item: T::ItemId,
+		lock_metadata: bool,
+		lock_attributes: bool,
```
[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/nfts/src/features/lock.rs:58]: You need to have a freezer role to make it non-transferable and call a `freeze_item()` method

[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/nfts/src/features/lock.rs:58]: > If we're being consistent with the above then this should take an `ItemConfig` rather than bools?

I made it this way in order to highlight that you can only modify 2 out of 3 settings with that method.
We can change it to ItemConfig and exclude `NonTransferable` if set. I just afraid that could create some confusion.
Wdyt?

File: frame/nfts/src/functions.rs (hunk 3 above)
Comment (jsidorenko) [reply frame/nfts/src/functions.rs]: Done

File: frame/nfts/src/features/lock.rs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::*;
+use frame_support::pallet_prelude::*;
+
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
```
[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/nfts/src/features/lock.rs:24]: done

File: frame/nfts/src/features/settings.rs (hunk 2 above)
Comment (jsidorenko) [reply frame/nfts/src/features/settings.rs:25]: done

File: frame/nfts/src/features/lock.rs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+		origin: T::AccountId,
+		collection: T::CollectionId,
+		lock_config: CollectionConfig,
+	) -> DispatchResult {
+		let details =
+			Collection::<T, I>::get(&collection).ok_or(Error::<T, I>::UnknownCollection)?;
+		ensure!(origin == details.freezer, Error::<T, I>::NoPermission);
+
+		CollectionConfigOf::<T, I>::try_mutate(collection, |maybe_config| {
+			let config = maybe_config.as_mut().ok_or(Error::<T, I>::UnknownCollection)?;
```
Comment (gilescope) [inline frame/nfts/src/features/lock.rs:35]: do we need to be able to distinguish this error from the one above? The more specific the error, the less support calls.

File: frame/nfts/src/features/settings.rs (hunk 2 above)
Comment (gilescope) [reply frame/nfts/src/features/settings.rs:25]: ah so these are part of the feature machinery rather than being a separate settings feature. Wondering if we should one module for features and another for the mechanism.

File: frame/nfts/src/functions.rs [hunk 8]
```diff
@@ -224,6 +240,13 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		ItemPriceOf::<T, I>::remove(&collection, &item);
 		PendingSwapOf::<T, I>::remove(&collection, &item);
 
+		// NOTE: if item's settings are not empty (e.g. item's metadata is locked)
+		// then we keep the record and don't remove it
+		let settings = Self::get_item_settings(&collection, &item)?;
+		if settings.is_empty() {
+			ItemConfigOf::<T, I>::remove(&collection, &item);
+		}
```
Comment (gilescope) [inline frame/nfts/src/functions.rs:248]: Is there an attack here where I mint and burn lots of items getting my deposit back but I rack up settings in the state?

File: frame/nfts/src/features/lock.rs (hunk 7 above)
Comment (jsidorenko) [reply frame/nfts/src/features/lock.rs:35]: In a lot of places (except the one you pointed out:) I removed the collection existence check and the first validation will be done over the config. So that if I pass a wrong id, I would get the `UnknownCollection` and not a `NoConfig` which would make more sense give the config would always exists together with collection

Comment (jsidorenko) [reply frame/nfts/src/features/lock.rs]: Though for that particular place that you highlighted I agree it would make sense to change the error message

File: frame/nfts/src/features/freeze.rs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+// limitations under the License.
+
+use crate::*;
+use frame_support::pallet_prelude::*;
+
+/// Freeze functions allow to make particular items non-transferable.
+/// It's also possible to revert the setting back.
+/// An origin must have a `Freezer` role in order to call those methods.
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	pub fn do_freeze_item(
```
Comment (ggwpez) [inline frame/nfts/src/features/freeze.rs:25]: The calls should be the only public functions on a pallet AFAIK. Everything else `pub(crate)`.

File: frame/nfts/src/functions.rs [hunk 10]
```diff
@@ -138,6 +145,9 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			CollectionAccount::<T, I>::remove(&collection_details.owner, &collection);
 			T::Currency::unreserve(&collection_details.owner, collection_details.total_deposit);
 			CollectionMaxSupply::<T, I>::remove(&collection);
+			CollectionConfigOf::<T, I>::remove(&collection);
+			#[allow(deprecated)]
+			ItemConfigOf::<T, I>::remove_prefix(&collection, None);
```
Comment (ggwpez) [inline frame/nfts/src/functions.rs:150]: The PoV Size of this is potentially unlimited. This is basically the same problem as https://github.com/paritytech/substrate/pull/12310 which favours a multi-step approach.

File: frame/nfts/src/impl_nonfungibles.rs [hunk 11]
```diff
@@ -78,26 +78,40 @@ impl<T: Config<I>, I: 'static> Inspect<<T as SystemConfig>::AccountId> for Palle
...
 	/// Create a `collection` of nonfungible items to be owned by `who` and managed by `admin`.
 	fn create_collection(
 		collection: &Self::CollectionId,
 		who: &T::AccountId,
 		admin: &T::AccountId,
+		settings: &CollectionSettings,
 	) -> DispatchResult {
+		let mut settings = *settings;
+		// FreeHolding could be set by calling the force_create() only
+		if settings.contains(CollectionSetting::FreeHolding) {
```
Comment (ggwpez) [inline frame/nfts/src/impl_nonfungibles.rs:106]: I think it would be better to have a wrapper type for clearer semantics:  
```rust
settings.is_enabled(…)
settings.is_disabled(…)
```

File: frame/nfts/src/lib.rs [hunk 12]
```diff
@@ -333,6 +314,23 @@ pub mod pallet {
 		OptionQuery,
 	>;
 
+	#[pallet::storage]
+	/// Config of a collection.
```
Comment (ggwpez) [inline frame/nfts/src/lib.rs:318]:
```suggestion
	/// Config of a collection.
	#[pallet::storage]
```

File: frame/nfts/src/lib.rs [hunk 13]
```diff
@@ -333,6 +314,23 @@ pub mod pallet {
...
+	pub(super) type CollectionConfigOf<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Blake2_128Concat, T::CollectionId, CollectionConfig, OptionQuery>;
+
+	#[pallet::storage]
+	/// Config of an item.
```
Comment (ggwpez) [inline frame/nfts/src/lib.rs:323]:
```suggestion
	/// Config of an item.
	#[pallet::storage]
```

File: frame/nfts/src/types.rs [hunk 14]
```diff
@@ -40,6 +43,12 @@ pub(super) type ItemTipOf<T, I = ()> = ItemTip<
 	BalanceOf<T, I>,
 >;
 
+pub trait Incrementable {
+	fn increment(&self) -> Self;
```
Comment (ggwpez) [inline frame/nfts/src/types.rs:47]:
```suggestion
	fn saturating_increment(&self) -> Self;
```

File: frame/nfts/src/types.rs [hunk 15]
```diff
@@ -40,6 +43,12 @@ pub(super) type ItemTipOf<T, I = ()> = ItemTip<
...
+	fn initial_value() -> Self;
+}
+impl_incrementable!(u8, u16, u32, u64, u128, i8, i16, i32, i64, i128);
```
Comment (ggwpez) [inline frame/nfts/src/types.rs:50]: You can probably implement this for all `Saturating + Zero` types.

File: frame/nfts/src/types.rs [hunk 16]
```diff
@@ -172,3 +171,89 @@ pub struct PriceWithDirection<Amount> {
 	/// A direction (send or receive).
 	pub(super) direction: PriceDirection,
 }
+
+// Support for up to 64 user-enabled features on a collection.
+#[bitflags]
```
Comment (ggwpez) [inline frame/nfts/src/types.rs:176]: I dont trust this macro. Could you add some test that checks that the values are indeed exclusive? So that they have no overlapping bits?

File: frame/nfts/src/types.rs [hunk 17]
```diff
@@ -172,3 +171,89 @@ pub struct PriceWithDirection<Amount> {
...
+	pub fn empty() -> Self {
+		Self(BitFlags::EMPTY)
+	}
+	pub fn values(&self) -> CollectionSettings {
+		self.0
+	}
+}
+impl_codec_bitflags!(CollectionConfig, u64, CollectionSetting);
+
+// Support for up to 64 user-enabled features on an item.
```
Comment (ggwpez) [inline frame/nfts/src/types.rs:205]:
```suggestion
/// Support for up to 64 user-enabled features on an item.
```

File: frame/support/src/traits/tokens/nonfungible_v2.rs [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+//! use.
+
+use super::nonfungibles_v2 as nonfungibles;
+use crate::{dispatch::DispatchResult, traits::Get};
+use codec::{Decode, Encode};
+use sp_runtime::TokenError;
+use sp_std::prelude::*;
+
+/// Trait for providing an interface to a read-only NFT-like set of items.
+pub trait Inspect<AccountId> {
```
Comment (ggwpez) [inline frame/support/src/traits/tokens/nonfungible_v2.rs:34]: How can I check if an item exists?

File: frame/support/src/traits/tokens/nonfungible_v2.rs [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+/// Convert a `fungibles` trait implementation into a `fungible` trait implementation by identifying
+/// a single item.
+pub struct ItemOf<
+	F: nonfungibles::Inspect<AccountId>,
+	A: Get<<F as nonfungibles::Inspect<AccountId>>::CollectionId>,
+	AccountId,
+>(sp_std::marker::PhantomData<(F, A, AccountId)>);
+
+impl<
+		F: nonfungibles::Inspect<AccountId>,
```
Comment (ggwpez) [inline frame/support/src/traits/tokens/nonfungible_v2.rs:126]: Writing it with a `where` clause makes it a bit more readable IHMO.

File: frame/support/src/traits/tokens/nonfungible_v2.rs [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+{
+	type ItemId = <F as nonfungibles::Inspect<AccountId>>::ItemId;
+	fn owner(item: &Self::ItemId) -> Option<AccountId> {
+		<F as nonfungibles::Inspect<AccountId>>::owner(&A::get(), item)
+	}
+	fn attribute(item: &Self::ItemId, key: &[u8]) -> Option<Vec<u8>> {
+		<F as nonfungibles::Inspect<AccountId>>::attribute(&A::get(), item, key)
+	}
+	fn typed_attribute<K: Encode, V: Decode>(item: &Self::ItemId, key: &K) -> Option<V> {
+		<F as nonfungibles::Inspect<AccountId>>::typed_attribute(&A::get(), item, key)
```
Comment (ggwpez) [inline frame/support/src/traits/tokens/nonfungible_v2.rs:139]: This only works if the collection ID is known at compile time, or?

File: frame/support/src/traits/tokens/nonfungibles_v2.rs [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+	fn create_collection(
+		collection: &Self::CollectionId,
+		who: &AccountId,
+		admin: &AccountId,
+		config: &CollectionConfig,
+	) -> DispatchResult;
+}
+
+/// Trait for providing the ability to destroy collections of nonfungible items.
+pub trait Destroy<AccountId>: Inspect<AccountId> {
```
Comment (ggwpez) [inline frame/support/src/traits/tokens/nonfungibles_v2.rs:136]: Why does a `Destroy` need to be an `Inspect`?

File: frame/support/src/traits/tokens/nonfungibles_v2.rs [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+		_key: &[u8],
+		_value: &[u8],
+	) -> DispatchResult {
+		Err(TokenError::Unsupported.into())
+	}
+
+	/// Attempt to set the strongly-typed attribute `value` of `collection`'s `key`.
+	///
+	/// By default this just attempts to use `set_attribute`.
+	fn set_typed_collection_attribute<K: Encode, V: Encode>(
```
Comment (ggwpez) [inline frame/support/src/traits/tokens/nonfungibles_v2.rs:224]: These traits probably need more a in-depth review than I have time to do rn.

**Review by ggwpez — APPROVED** [review]
Look good from afar, but should be fine since it merges into an internal branch.

File: frame/nfts/src/features/settings.rs (hunk 2 above)
Comment (jsidorenko) [reply frame/nfts/src/features/settings.rs:25]: @gilescope where would you put them?

File: frame/nfts/src/features/atomic_swap.rs [hunk 23]
```diff
@@ -31,6 +31,10 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		maybe_price: Option<PriceWithDirection<ItemPrice<T, I>>>,
 		duration: <T as SystemConfig>::BlockNumber,
 	) -> DispatchResult {
+		ensure!(
+			Self::is_pallet_feature_disabled(PalletFeature::NoSwaps),
```
[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/nfts/src/features/atomic_swap.rs:35]: It's because by default the flags are empty and we only store what is disabled. This applies to all the settings: pallet's, collection's, item's.

File: frame/support/src/traits/tokens/nonfungibles_v2.rs (hunk 21 above)
Comment (jsidorenko) [reply frame/support/src/traits/tokens/nonfungibles_v2.rs:136]: Idk, it's a copy of the existing `frame/support/src/traits/tokens/nonfungibles.rs` I only added support for feature flags

File: frame/nfts/src/features/freeze.rs (hunk 9 above)
Comment (jsidorenko) [reply frame/nfts/src/features/freeze.rs:25]: What if I want to call some function from another pallet? Afaik, the main usage of the uniques pallet is to provide the basic implementation to deal with NFTs and all the custom logic will be placed in other pallets.

File: frame/nfts/src/features/lock.rs [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+/// Lock functions allow to lock collection/items metadata and attributes.
+/// Additionally, it's possible to make all collection items non-transferable.
+/// Those settings are irreversible.
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	pub fn do_lock_collection(
+		origin: T::AccountId,
+		collection: T::CollectionId,
+		lock_config: CollectionConfig,
+	) -> DispatchResult {
+		let details =
```
[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/nfts/src/features/lock.rs:30]: Yep, this allows to change different collection-level settings, I assume that method should always be accessible

File: frame/nfts/src/features/freeze.rs (hunk 9 above)
Comment (gilescope) [reply frame/nfts/src/features/freeze.rs:25]: Any other pallet building on it should be able to go through the public api.

File: frame/nfts/src/features/lock.rs (hunk 24 above)
[the remark this answers is not in the document]
Comment (gilescope) [reply frame/nfts/src/features/lock.rs:30]: we could state that in the docs here so people know why it should not be turned into a feature?

File: frame/nfts/src/features/settings.rs (hunk 2 above)
Comment (gilescope) [reply frame/nfts/src/features/settings.rs:25]: Here's fine for now. We can revisit locations when we have another pallet that needs the same code.

File: frame/nfts/src/functions.rs (hunk 10 above)
Comment (gilescope) [reply frame/nfts/src/functions.rs:150]: Maybe one of the standard Error enums should be `Error::IncompletePleaseCallAgain` ?

File: frame/nfts/src/lib.rs [hunk 25]
```diff
@@ -499,8 +499,10 @@ pub mod pallet {
 		WrongOwner,
 		/// Invalid witness data given.
 		BadWitness,
```
Comment (gilescope) [inline frame/nfts/src/lib.rs:501]:
```suggestion
		/// The witness data given does not match the current state of the chain.
		BadWitness,
```

File: frame/nfts/src/lib.rs [hunk 26]
```diff
@@ -499,8 +499,10 @@ pub mod pallet {
...
-		/// The item ID is already taken.
-		InUse,
+		/// Collection ID is already taken.
+		CollectionIdInUse,
+		/// Items within that collection are non-transferable.
+		ItemsNotTransferable,
 		/// The item or collection is frozen.
```
Comment (gilescope) [inline frame/nfts/src/lib.rs:506]: would be good to say what 'frozen' means here. Be good to state the difference between frozen and locked.

File: frame/nfts/src/lib.rs [hunk 27]
```diff
@@ -1262,20 +1272,35 @@ pub mod pallet {
 			key: BoundedVec<u8, T::KeyLimit>,
 			value: BoundedVec<u8, T::ValueLimit>,
 		) -> DispatchResult {
+			ensure!(
+				Self::is_pallet_feature_disabled(PalletFeature::NoAttributes),
```
Comment (gilescope) [inline frame/nfts/src/lib.rs:1276]: Shouldn't features be always positive like rust features? I.e. the feature should be Attributes - it could be a default feature.

File: frame/nfts/src/tests.rs [hunk 28]
```diff
@@ -398,53 +426,57 @@ fn set_item_metadata_should_work() {
...
-			Error::<Test>::NoPermission
+			Error::<Test>::NoPermission,
 		);
 		assert_noop!(
 			Nfts::clear_metadata(RuntimeOrigin::signed(1), 1, 42),
-			Error::<Test>::UnknownCollection
+			Error::<Test>::UnknownCollection,
 		);
-		assert_ok!(Nfts::clear_metadata(RuntimeOrigin::signed(1), 0, 42));
+		assert_ok!(Nfts::clear_metadata(RuntimeOrigin::root(), 0, 42));
```
Comment (gilescope) [inline frame/nfts/src/tests.rs:479]: Why does this now have to be root?

File: frame/nfts/src/types.rs (hunk 16 above)
Comment (gilescope) [reply frame/nfts/src/types.rs:176]: this is a good point, we should axe enumflags2 crate in favour of bitflags crate as that's more commonly used in substrate. less deps is better security.

File: frame/nfts/src/features/macros.rs (hunk 1 above)
Comment (gilescope) [reply frame/nfts/src/features/macros.rs:71]: Hmm, let's wait till another pallet uses features.

File: frame/nfts/src/lib.rs (hunk 27 above)
Comment (gilescope) [reply frame/nfts/src/lib.rs:1276]: This is the same thing as Oliver's point above.

File: frame/nfts/src/features/atomic_swap.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply frame/nfts/src/features/atomic_swap.rs:35]: Why? Does not matter whether you use `0...` or `1...` as default.

File: frame/nfts/src/features/freeze.rs (hunk 9 above)
Comment (ggwpez) [reply frame/nfts/src/features/freeze.rs:25]: Through traits

File: frame/nfts/src/features/macros.rs [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+macro_rules! impl_incrementable {
+	($($type:ty),+) => {
+		$(
+			impl Incrementable for $type {
+				fn increment(&self) -> Self {
+					self.saturating_add(1)
```
[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/nfts/src/features/macros.rs]: done

File: frame/nfts/src/features/freeze.rs [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::*;
+use frame_support::pallet_prelude::*;
+
+/// Freeze functions allow to make particular items non-transferable.
```
[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/nfts/src/features/freeze.rs]: done

File: frame/nfts/src/features/atomic_swap.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/nfts/src/features/atomic_swap.rs]: done

File: bin/node/runtime/src/lib.rs [hunk 31]
```diff
@@ -1497,6 +1498,10 @@ impl pallet_uniques::Config for Runtime {
 	type Locker = ();
 }
 
+parameter_types! {
+	pub Features: PalletFeatures = PalletFeatures::empty();
```
[the remark this answers is not in the document]
Comment (jsidorenko) [reply bin/node/runtime/src/lib.rs]: done

File: frame/nfts/src/features/freeze.rs (hunk 9 above)
Comment (jsidorenko) [reply frame/nfts/src/features/freeze.rs]: Done

File: frame/nfts/src/types.rs (hunk 14 above)
Comment (jsidorenko) [reply frame/nfts/src/types.rs:47]: Let's do that during the final PR to master

File: frame/nfts/src/types.rs (hunk 15 above)
Comment (jsidorenko) [reply frame/nfts/src/types.rs:50]: same as above

File: frame/nfts/src/types.rs (hunk 16 above)
Comment (jsidorenko) [reply frame/nfts/src/types.rs:176]: We already use that macro in the identity pallet, it's not new here

File: frame/nfts/src/lib.rs [hunk 32]
```diff
@@ -512,7 +514,15 @@ pub mod pallet {
 		/// The named owner has not signed ownership of the collection is acceptable.
```
[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/nfts/src/lib.rs:515]: `/// The named owner has not signed ownership acceptance of the collection.` Is that better? :)

File: frame/nfts/src/lib.rs (hunk 25 above)
Comment (jsidorenko) [reply frame/nfts/src/lib.rs]: Done

File: frame/support/src/traits/tokens/nonfungibles_v2.rs (hunk 22 above)
Comment (jsidorenko) [reply frame/support/src/traits/tokens/nonfungibles_v2.rs:224]: Since I only copy+pasted them from v1, let's work on them in a separate PR, wdyt?

File: frame/nfts/src/tests.rs (hunk 28 above)
Comment (jsidorenko) [reply frame/nfts/src/tests.rs:479]: When the metadata is locked, only the root has rights to clear it. There is a check for that a couple of line above:
```
		assert_noop!(
			Nfts::clear_metadata(RuntimeOrigin::signed(1), 0, 42),
			Error::<Test>::LockedItemMetadata,
		);
```

**Review by ggwpez — APPROVED** [review]

File: frame/nfts/src/lib.rs (hunk 26 above)
Comment (jsidorenko) [reply frame/nfts/src/lib.rs:501]: Done

**Comment by jsidorenko** [comment]
@gilescope @ggwpez Guys, I re-wrote the whole implementation, swapped negative setting names with positive and added different helpers to make the code more readable. Plus I renamed freeze/lock/thaw methods to remove the ambiguity.
Could you pls review this PR again?

**Review by gilescope — APPROVED** [review]
Excellent, most happy.

File: frame/nfts/src/functions.rs [hunk 33]
```diff
@@ -36,12 +36,22 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	) -> DispatchResult {
 		let collection_details =
 			Collection::<T, I>::get(&collection).ok_or(Error::<T, I>::UnknownCollection)?;
-		ensure!(!collection_details.is_frozen, Error::<T, I>::Frozen);
-		ensure!(!T::Locker::is_locked(collection, item), Error::<T, I>::Locked);
+		ensure!(!T::Locker::is_locked(collection, item), Error::<T, I>::ItemLocked);
```
Comment (ggwpez) [inline frame/nfts/src/functions.rs:39]: The `Frozen` error enum constant is not used anymore.

File: frame/nfts/src/functions.rs (hunk 10 above)
Comment (ggwpez) [reply frame/nfts/src/functions.rs:152]: The witness should probably contain an upper limit to bound the weight.

File: frame/nfts/src/lib.rs [hunk 34]
```diff
@@ -565,19 +576,29 @@ pub mod pallet {
...
 				NextCollectionId::<T, I>::get().unwrap_or(T::CollectionId::initial_value());
 
 			let owner = T::CreateOrigin::ensure_origin(origin, &collection)?;
 			let admin = T::Lookup::lookup(admin)?;
 
+			let mut config = config;
+			// DepositRequired could be disabled by calling the force_create() only
+			if config.has_disabled_setting(CollectionSetting::DepositRequired) {
+				config.enable_setting(CollectionSetting::DepositRequired);
+			}
```
Comment (ggwpez) [inline frame/nfts/src/lib.rs:594]:
```suggestion
config.enable_setting(CollectionSetting::DepositRequired);
```
I think it would be better to error here though. Otherwise the creator will wonder why it is enabled.

File: frame/nfts/src/types.rs [hunk 35]
```diff
@@ -172,3 +176,128 @@ pub struct PriceWithDirection<Amount> {
...
+}
+pub(super) type CollectionSettings = BitFlags<CollectionSetting>;
+
+/// Wrapper type for `CollectionSettings` that implements `Codec`.
+#[derive(Clone, Copy, PartialEq, Eq, Default, RuntimeDebug)]
+pub struct CollectionConfig(pub CollectionSettings);
+
+impl CollectionConfig {
+	pub fn all_settings_enabled() -> Self {
+		Self(BitFlags::EMPTY)
```
Comment (ggwpez) [inline frame/nfts/src/types.rs:202]: Normally bitflags are used in a way that a 1 signals enabled and 0 disabled.  
If you use `EMPTY` as enabled, then this will require a storage migration if you add a new setting, unless you implicitly want to force-enable it onto all deployed nfts+collections.  
1 = enabled is more defensive.

File: frame/nfts/src/benchmarking.rs [hunk 36]
```diff
@@ -146,25 +145,19 @@ benchmarks_instance_pallet! {
...
 		let caller_lookup = T::Lookup::unlookup(caller.clone());
-	}: _(SystemOrigin::Root, caller_lookup, true)
+	}: _(SystemOrigin::Root, caller_lookup, CollectionConfig::all_settings_enabled())
 	verify {
 		assert_last_event::<T, I>(Event::ForceCreated { collection: T::Helper::collection(0), owner: caller }.into());
 	}
 
 	destroy {
 		let n in 0 .. 1_000;
-		let m in 0 .. 1_000;
```
Comment (ggwpez) [inline frame/nfts/src/benchmarking.rs:156]: In the weight file it looks like these components are used. Why did you remove them?

File: frame/nfts/src/types.rs (hunk 35 above)
Comment (jsidorenko) [reply frame/nfts/src/types.rs:202]: You set 1 (true) for the disabled things, so if nothing is disabled, then there is no need to store anything.
Why would you need to run a migration for the new settings? If you add smth new, it will be enabled by default, until you explicitly disable it

File: frame/nfts/src/benchmarking.rs (hunk 36 above)
Comment (jsidorenko) [reply frame/nfts/src/benchmarking.rs:156]: Because previously it was possible to create metadata and attributes for non-existing records and that benchmark worked. After I added a check whether the record exists, that benchmark started to fail when we try to add metadata (or an attribute) in separate loop

File: frame/nfts/src/features/lock.rs [hunk 37]
```diff
@@ -0,0 +1,120 @@
...
+
+			if lock_config.has_disabled_setting(CollectionSetting::TransferableItems) {
+				config.disable_setting(CollectionSetting::TransferableItems);
+			}
+			if lock_config.has_disabled_setting(CollectionSetting::UnlockedMetadata) {
+				config.disable_setting(CollectionSetting::UnlockedMetadata);
+			}
+			if lock_config.has_disabled_setting(CollectionSetting::UnlockedAttributes) {
+				config.disable_setting(CollectionSetting::UnlockedAttributes);
+			}
```
[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/nfts/src/features/lock.rs:42]: I'm checking only those 3 settings since the last one (`DepositRequired`) can be set by the root only. In the future, I expect more settings with custom business logic or gated control

File: frame/nfts/src/lib.rs (hunk 34 above)
Comment (jsidorenko) [reply frame/nfts/src/lib.rs]: Nice suggestion!

File: frame/nfts/src/features/lock.rs (hunk 37 above)
[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/nfts/src/features/lock.rs:42]: > With bit flags something config &= !lock_config should work. Or more Rusty and less Cy: config -= lock_config.

Wouldn't that significantly reduce the code readability? Giles wrote above that we should write a more readable code, that's why I added `has_disabled_setting()`, `disable_setting()` and other helpers
